src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,943 @@
+  printf("pixel representation = %u\n",
+         dcm_frame_get_pixel_representation(frame));
+  printf("planar configuration = %u\n",
+         dcm_frame_get_planar_configuration(frame));
+  printf("photometric interpretation = %s\n",
+         dcm_frame_get_photometric_interpretation(frame));
+  printf("transfer syntax uid = %s\n",
+         dcm_frame_get_transfer_syntax_uid(frame));
+}
+#endif /*DEBUG*/
+
+static bool dicom_detect(const char *filename,
+                         struct _openslide_tifflike *tl, 
+                         GError **err) {
+  // reject TIFFs
+  if (tl) {
+    g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                "Is a TIFF file");
+    return false;
+  }
+
+  // verify existence
+  GError *tmp_err = NULL;
+  if (!_openslide_fexists(filename, &tmp_err)) {
+    if (tmp_err != NULL) {
+      g_propagate_prefixed_error(err, tmp_err, "Testing whether file exists: ");
+    } else {
+      g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                  "File does not exist");
+    }
+    return false;
+  }
+
+  // should be able to open as a DICOM
+  DcmError *dcm_error = NULL;
+  g_autoptr(DcmFilehandle) filehandle = 
+    dcm_filehandle_create_from_file(&dcm_error, filename);
+  if (filehandle == NULL) {
+    set_gerror_from_dcm_error(err, &dcm_error);
+    return false;
+  }
+
+  // should be able to open as a DICOM
+  g_autoptr(DcmDataSet) meta = 
+    dcm_filehandle_read_file_meta(&dcm_error, filehandle);
+  if (!meta) {
+    set_gerror_from_dcm_error(err, &dcm_error);
+    return false;
+  }
+
+  return true;
+}
+
+static void dicom_file_destroy(struct dicom_file *f) {
+  FREEF(dcm_filehandle_destroy, f->filehandle);
+  FREEF(dcm_dataset_destroy, f->meta);
+  FREEF(dcm_dataset_destroy, f->metadata);
+  FREEF(dcm_bot_destroy, f->bot);
+  g_mutex_clear(&f->lock);
+  g_free(f->filename);
+  g_slice_free(struct dicom_file, f);
+}
+
+static bool get_tag_int(DcmDataSet *dataset, const char *keyword, int *result) {
+  uint32_t tag = dcm_dict_tag_from_keyword(keyword);
+  DcmElement *element = dcm_dataset_get(NULL, dataset, tag);
+  int64_t value;
+  if (!dcm_element_get_value_integer(NULL, element, 0, &value)) {
+    return false;
+  }
+  *result = value;
+  return true;
+}
+
+static bool get_tag_str(DcmDataSet *dataset, 
+                        const char *keyword, 
+                        int index, 
+                        const char **result) {
+  uint32_t tag = dcm_dict_tag_from_keyword(keyword);
+  DcmElement *element = dcm_dataset_get(NULL, dataset, tag);
+  return dcm_element_get_value_string(NULL, element, index, result);
+}
+
+static struct dicom_file *dicom_file_new(char *filename, GError **err) {
+  struct dicom_file *f = g_slice_new0(struct dicom_file);
+  f->filename = g_strdup(filename);
+
+  // should be able to open as a DICOM
+  DcmError *dcm_error = NULL;
+  f->filehandle = dcm_filehandle_create_from_file(&dcm_error, f->filename);
```

Comment:
We'll need to indirect libdicom's file accesses to go through OpenSlide's VFS.  Note that (for better or for worse) OpenSlide generally does not hold `FILE` pointers open between calls into the OpenSlide API.

Reply:
Implemented with https://github.com/openslide/openslide/pull/431/commits/3d2f50d72775d5e5ab0360cd2edff99922c56ce9